seatunnel-config/src/main/java/com/typesafe/config/impl/Path.java

```diff
@@ -0,0 +1,254 @@
+            count += 1;
+            p = p.remainder;
+        }
+        return count;
+    }
+
+    Path subPath(int removeFromFront) {
+        int count = removeFromFront;
+        Path p = this;
+        while (p != null && count > 0) {
+            count -= 1;
+            p = p.remainder;
+        }
+        return p;
+    }
+
+    Path subPath(int firstIndex, int lastIndex) {
+        if (lastIndex < firstIndex) {
+            throw new ConfigException.BugOrBroken("bad call to subPath");
+        }
+
+        Path from = subPath(firstIndex);
+        PathBuilder pb = new PathBuilder();
+        int count = lastIndex - firstIndex;
+        while (count > 0) {
+            count -= 1;
+            pb.appendKey(from.first());
+            from = from.remainder();
+            if (from == null) {
+                throw new ConfigException.BugOrBroken("subPath lastIndex out of range " + lastIndex);
+            }
+        }
+        return pb.result();
+    }
+
+    boolean startsWith(Path other) {
+        Path myRemainder = this;
+        Path otherRemainder = other;
+        if (otherRemainder.length() <= myRemainder.length()) {
+            while (otherRemainder != null) {
+                if (!otherRemainder.first().equals(myRemainder.first())) {
+                    return false;
+                }
+                myRemainder = myRemainder.remainder();
+                otherRemainder = otherRemainder.remainder();
+            }
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other instanceof Path) {
+            Path that = (Path) other;
+            return this.first.equals(that.first)
+                    && ConfigImplUtil.equalsHandlingNull(this.remainder,
+                    that.remainder);
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return 41 * (41 + first.hashCode())
+                + (remainder == null ? 0 : remainder.hashCode());
+    }
+
+    // this doesn't have a very precise meaning, just to reduce
+    // noise from quotes in the rendered path for average cases
+    static boolean hasFunkyChars(String s) {
+        int length = s.length();
+
+        if (length == 0) {
+            return false;
+        }
+
+        for (int i = 0; i < length; ++i) {
+            char c = s.charAt(i);
+
+            if (Character.isLetterOrDigit(c) || c == '-' || c == '_' || c == '.') {
+                continue;
+            } else {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void appendToStringBuilder(StringBuilder sb) {
+        if (hasFunkyChars(first) || first.isEmpty()) {
+            sb.append(ConfigImplUtil.renderJsonString(first));
+        } else {
+            sb.append(first);
+        }
+        if (remainder != null) {
+            sb.append(ConfigParseOptions.PATH_TOKEN_SEPARATOR);
```

Comment:
cc @leo65535  FYI https://github.com/apache/incubator-seatunnel/blob/v2.0.4/waterdrop-config/src/main/java/io/github/interestinglab/waterdrop/config/impl/Path.java#L203

Reply:
thanks, will check it.